ArduCopter/mode_auto.cpp

```diff
@@ -1165,21 +1165,192 @@ void ModeAuto::payload_place_run()
+            break;
+        } else if (now - nav_payload_place.descent_established_timestamp < descend_throttle_calibrate_time) {
+            // record minimum thrust for descend_throttle_calibrate_time
+            nav_payload_place.descend_throttle_level = MIN(nav_payload_place.descend_throttle_level, current_throttle_level);
+            nav_payload_place.place_start_timestamp = now;
+            break;
+        } else if (current_throttle_level > g2.pldp_throttle_placed_fraction * nav_payload_place.descend_throttle_level) {
+            // thrust is above minimum threshold
+            nav_payload_place.place_start_timestamp = now;
+            break;
+        } else if (is_positive(g2.pldp_range_finder_minimum)) {
+            if (!copter.rangefinder_state.enabled) {
+                // abort payload place because rangefinder is not enabled
+                nav_payload_place.state = PayloadPlaceStateType_Ascent_Start;
+                gcs().send_text(MAV_SEVERITY_WARNING, "PayloadPlace: PLDP_RNG_MIN set and rangefinder not enabled");
+                break;
+            } else if (copter.rangefinder_alt_ok() && (copter.rangefinder_state.glitch_count == 0) && (copter.rangefinder_state.alt_cm > g2.pldp_range_finder_minimum * 100.0)) {
+                // range finder altitude is above minimum
+                nav_payload_place.place_start_timestamp = now;
+                break;
+            }
+        }
+
+        // If we get here:
+        // 1. we have reached decent velocity
+        // 2. measured the thrust level required for decent
+        // 3. detected that our thrust requirements have reduced
+        // 4. rangefinder range has dropped below minimum if set
+        // 5. place_start_timestamp has been initialised
+
+        // payload touchdown must be detected for 0.5 seconds
+
+        if (now - nav_payload_place.place_start_timestamp > placed_time) {
+            nav_payload_place.state = PayloadPlaceStateType_Release;
+            gcs().send_text(MAV_SEVERITY_INFO, "PayloadPlace: payload release thrust threshold: %f", static_cast<double>(g2.pldp_throttle_placed_fraction * nav_payload_place.descend_throttle_level));
+        }
+        break;
+
+    case PayloadPlaceStateType_Release:
+        // Reinitialise vertical position controller to remove discontinuity due to touch down of payload
+        pos_control->init_z_controller_no_descent();
+#if AP_GRIPPER_ENABLED == ENABLED
+        if (g2.gripper.valid()) {
+            gcs().send_text(MAV_SEVERITY_INFO, "PayloadPlace: Releasing the gripper");
+            g2.gripper.release();
+            nav_payload_place.state = PayloadPlaceStateType_Releasing;
+        } else {
+            gcs().send_text(MAV_SEVERITY_INFO, "PayloadPlace: Gripper not valid");
+            nav_payload_place.state = PayloadPlaceStateType_Ascent_Start;
+        }
+#else
+        gcs().send_text(MAV_SEVERITY_INFO, "Gripper code disabled");
```

Comment:
I think maybe this "Gripper code disabled" could be "PayloadPlace: Gripper disabled" because it is payload place that is sending the message.